339 ssd ec2 instance type

nikki-t opened the pull request:
## Purpose

Experiment with using EC2 types with attached storage for better performance when pulling docker images

## Proposed Changes

- [ADD] - `user-data` script that mounts instance store SSDs to underlying `emptyDir` directory found at `/var/lib/kubelet/pods`
- [CHANGE] - Modify `cwl_dag.py` to mount `/data/` and `/var/lib/docker` to `emptyDir`
- [CHANGE] - Modify `cwl_dag.py` to include an input parameter specifying use of instance store
- [CHANGE] - Update `variables.tf` to include `m5ad` instance family.

## Issues

- #339 - [New Feature]: Experiment with using EC2 types with attached storage for better performance

## Testing

Ran in `unity-venue-dev` on `unity-nikki-2` deployment. `nvme1n1` is the instance store SSD.

- Confirmed node mounted `/var/lib/kubelet/pods`

```bash
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        7.7G     0  7.7G   0% /dev
tmpfs           7.7G     0  7.7G   0% /dev/shm
tmpfs           7.7G  2.0M  7.7G   1% /run
tmpfs           7.7G     0  7.7G   0% /sys/fs/cgroup
/dev/nvme0n1p1  300G  6.5G  294G   3% /
tmpfs           7.7G  4.0K  7.7G   1% /tmp
/dev/nvme1n1    140G  1.9G  138G   2% /var/lib/kubelet/pods
tmpfs           768M   12K  768M   1% /var/lib/kubelet/pods/db75e74e-29a0-491b-8677-9d9883698c8a/volumes/kubernetes.io~projected/kube-api-access-67njd
shm              64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/321097646a8818ed629031d9a082844699ae043e0607b5ed57249406d88b01a0/shm
shm              64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/7737fa11d962e483464bf642b998219f73198b8e1a4c30d20d56d7ae10cf977f/shm
shm              64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/0e7d4c5e6b9bfea4bd87e17a5da46c554c4b315c8c99b4bd6a26d6bab037573b/shm
shm              64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/f816fc1768b68e19bd63894047a609bd35af8255f196d1dad7d10c333c730d26/shm
shm              64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/90b756f456282e302e13a493cb87930ab92195bcb4dd9c32580fc42a351d18b6/shm
```

- Confirmed pod mounted `/data/` and `/var/lib/docker`

```bash
Filesystem                Size      Used Available Use% Mounted on
overlay                 300.0G      5.9G    294.1G   2% /
tmpfs                    64.0M         0     64.0M   0% /dev
tmpfs                     7.6G         0      7.6G   0% /sys/fs/cgroup
/dev/nvme1n1            139.6G      2.6G    137.0G   2% /data
/dev/nvme1n1            139.6G      2.6G    137.0G   2% /etc/hosts
/dev/nvme1n1            139.6G      2.6G    137.0G   2% /dev/termination-log
/dev/nvme0n1p1          300.0G      5.9G    294.1G   2% /etc/hostname
/dev/nvme0n1p1          300.0G      5.9G    294.1G   2% /etc/resolv.conf
shm                      64.0M         0     64.0M   0% /dev/shm
/dev/nvme1n1            139.6G      2.6G    137.0G   2% /var/lib/docker
tmpfs                    14.3G     12.0K     14.3G   0% /run/secrets/kubernetes.io/serviceaccount
tmpfs                    14.3G      4.0K     14.3G   0% /run/secrets/eks.amazonaws.com/serviceaccount
```

Ran EMIT workflow which completed successfully: https://www.dev.mdps.mcp.nasa.gov:4443/unity-nikki-2/dev/sps/dags/cwl_dag/grid?dag_run_id=manual__2025-03-14T18%3A44%3A20%2B00%3A00&task_id=cwl_task&tab=logs

LucaCinquini reviewed (changes requested):
Nikki, amazing job here.
I would recommend the following change though: instead of having an additional parameter to use instance store, which overrides the EC2 selection, we should just let the user choose m5ad.xlarge, and perhaps a few other types, which will cause that to happen. In the UI, the description for "m5ad.xlarge" would contain something like "General Type with SSD storage attached". WDYT?

nikki-t commented:
@LucaCinquini - I updated the CWL DAG and utilities script to include `m5ad` instance types and updated the description to include "General Purpose with SSD storage":

![Screenshot 2025-03-17 at 09 16 52](https://github.com/user-attachments/assets/f73c75e2-74cf-497d-b717-48894ecb5071)

I also ran on the m5ad.4xlarge and made sure the `md0` RAID array was mounted correctly and included both SSD disks on the node and pod.

```bash
Filesystem                Size      Used Available Use% Mounted on
overlay                 300.0G      6.3G    293.6G   2% /
tmpfs                    64.0M         0     64.0M   0% /dev
tmpfs                    30.8G         0     30.8G   0% /sys/fs/cgroup
/dev/md0                549.6G     84.7M    521.6G   0% /data
/dev/md0                549.6G     84.7M    521.6G   0% /etc/hosts
/dev/md0                549.6G     84.7M    521.6G   0% /dev/termination-log
/dev/nvme0n1p1          300.0G      6.3G    293.6G   2% /etc/hostname
/dev/nvme0n1p1          300.0G      6.3G    293.6G   2% /etc/resolv.conf
shm                      64.0M         0     64.0M   0% /dev/shm
/dev/md0                549.6G     84.7M    521.6G   0% /var/lib/docker
tmpfs                    58.8G     12.0K     58.8G   0% /run/secrets/kubernetes.io/serviceaccount
tmpfs                    58.8G      4.0K     58.8G   0% /run/secrets/eks.amazonaws.com/serviceaccount
overlay                 549.6G     84.7M    521.6G   0% /var/lib/docker/overlay2/2d338bb29da3912d3e7ca771a24ada205e892a1551623d843cbbe4ae70d32e5c/merged
```

LucaCinquini reviewed (approved):
All looks good, thanks.
